src/crimson/os/seastore/linked_tree_node.h

```diff
@@ -0,0 +1,1002 @@
+    if (mut_iter != me.mutation_pendings.end()) {
+      assert(copy_dests_by_trans.find(t.get_trans_id()) ==
+	copy_dests_by_trans.end());
+      return static_cast<T*>(&(*mut_iter));
+    }
+    auto iter = copy_dests_by_trans.find(
+      t.get_trans_id(), trans_spec_view_t::cmp_t());
+    ceph_assert(iter != copy_dests_by_trans.end());
+    auto &copy_dests = static_cast<copy_dests_t&>(*iter);
+    auto it = copy_dests.dests_by_key.lower_bound(key);
+    if (it == copy_dests.dests_by_key.end() || (*it)->range.begin > key) {
+      ceph_assert(it != copy_dests.dests_by_key.begin());
+      --it;
+    }
+    ceph_assert((*it)->range.begin <= key && key < (*it)->range.end);
+    return *it;
+  }
+
+  template <typename ChildT>
+  get_child_ret_t<T, ChildT> get_child(
+    Transaction &t,
+    ExtentTransViewRetriever &etvr,
+    btreenode_pos_t pos,
+    node_key_t key)
+  {
+    auto &me = down_cast();
+    assert(children.capacity());
+    assert(key == down_cast().iter_idx(pos).get_key());
+    auto child = children[pos];
+    ceph_assert(!is_reserved_ptr(child));
+    if (is_valid_child_ptr(child)) {
+      return etvr.get_extent_viewable_by_trans<ChildT>(
+	t, static_cast<ChildT*>(child));
+    } else if (me.is_pending()) {
+      auto &sparent = me.get_stable_for_key(key);
+      auto spos = sparent.lower_bound(key).get_offset();
+      auto child = sparent.children[spos];
+      if (is_valid_child_ptr(child)) {
+	return etvr.get_extent_viewable_by_trans<ChildT>(
+	  t, static_cast<ChildT*>(child));
+      } else {
+        etvr.account_absent_access(t.get_src());
+	return child_pos_t<T>(&sparent, spos);
+      }
+    } else {
+      etvr.account_absent_access(t.get_src());
+      return child_pos_t<T>(&me, pos);
+    }
+  }
+
+  void link_child(BaseChildNode<T, node_key_t>* child, btreenode_pos_t pos) {
+    auto &me = down_cast();
+    assert(pos < me.get_size());
+    assert(child);
+    ceph_assert(!me.is_pending());
+    assert(child->valid() && !child->pending());
+    assert(!children[pos]);
+    ceph_assert(is_valid_child_ptr(child));
+    update_child_ptr(pos, child);
+  }
+
+  void insert_child_ptr(btreenode_pos_t offset, BaseChildNode<T, node_key_t>* child) {
+    assert(child);
+    auto raw_children = children.data();
+    std::memmove(
+      &raw_children[offset + 1],
+      &raw_children[offset],
+      (num_children - offset) * sizeof(BaseChildNode<T, node_key_t>*));
+    children[offset] = child;
+    if (!is_reserved_ptr(child)) {
+      set_child_ptracker(child);
+    }
+    num_children++;
+  }
+
+  void update_child_ptr(btreenode_pos_t pos, BaseChildNode<T, node_key_t>* child) {
+    assert(child);
+    assert(pos < down_cast().get_size());
+    assert(pos < num_children);
+    children[pos] = child;
+    set_child_ptracker(child);
+  }
+
+  btreenode_pos_t get_num_children() const {
+    return num_children;
+  }
+
+  void sync_num_children() {
+    auto &me = down_cast();
+    num_children = me.get_size();
```

Comment:
Why do we need to worry about sync and update an independent `num_children`? What can be wrong if directly query `down_cast().get_size()` instead?

Reply:
I thought that there were times that we must modify the layout, which changes the value of `me.get_size()`, before touching the `children` vector. But it doesn't seem necessary to me now, will check on it again.

Reply:
@cyx1231st The `copy_source/dest` related operations rely on the correct node layout, so I moved the `copy_source/dest` related operations out of `split/merge/balance_child_ptrs` which are called before modifying the layout, and do the `copy_source/dest` operations after the modification of the node layout, which renderred the independent `num_children` unnecessary. Please take a look, again. Thanks:-)